bittensor/_neuron/text/multitron_validator/neuron_impl.py

```diff
@@ -0,0 +1,484 @@
+
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+""" Base neuron version 1.
+
+Example:
+    $ import neurons
+    $ neurons.text.base_neuron_v1().run()
+
+"""
+import os
+import pandas
+from pandas.core.frame import DataFrame
+import bittensor
+import math
+import torch
+import torch.multiprocessing as mp
+from torch.nn.parallel import DistributedDataParallel as DDP
+import torch.distributed as dist
+import traceback
+import sys
+import wandb
+from termcolor import colored
+from qqdm import qqdm, format_str
+from loguru import logger
+
+from bittensor._metagraph import metagraph
+logger = logger.opt(colors=True)
+
+from types import SimpleNamespace
+from torch.nn.utils import clip_grad_norm_
+import torch.nn as nn
+from functools import partial
+
+import torch.nn.functional as F
+from torch.multiprocessing import Manager
+import pickle
+from os import path
+import random
+import time
+import datetime
+
+class DDPNeuronTrain:
+    def __init__( self, config: 'bittensor.config', nucleus: 'Nucleus', wallet: 'bittensor.wallet'):
+        r""" Initializes the neuron with the passed config.
+        """
+        self.config = config
+        self.wallet = wallet
+        self.world_size = config.neuron.world_size
+        self.nucleus = nucleus
+        self.stats = SimpleNamespace(
+            global_step = 0,
+            last_sync_block = 0,
+            epoch_data_size = 0,
+            epoch_sync_count = 0,
+            local_target_epoch_loss = math.inf,
+            distillation_epoch_loss = math.inf,
+            remote_target_epoch_loss = math.inf,
+            local_epoch_acc = 0,
+            best_epoch_loss = math.inf,
+            scores = {},
+            ema_scores = torch.nn.Parameter(torch.zeros(self.config.nucleus.max_n), requires_grad = False)
+        )
+        # ---- Decay factor for fisher ema score 
+        self.fisher_ema_decay = 0.995
+
+    def stop( self ):
+        r""" Stop the dendrite and dataset
+        """
+        del self.dendrite
+        self.dataset.close()
+    
+    def init_process(self, rank):
+        r""" For each process, anchor them to the process group 
+        so that they know how to communication with each other.
+
+        Args:
+            rank (int):
+                rank (id) of the process.
+        """
+        os.environ['MASTER_ADDR'] = self.config.neuron.address
```

Comment:
Can we make it more specific? Maybe BIT_MASTER_ADDR?

Reply:
This one cannot do, cause MASTER_ADDR is the env variable that DDP listens to